### DISCUSSION TITLE
FAST.Farm Modeling Question

### DISCUSSION THREAD

**Discussion by Tommi1198** [OpenFAST/openfast#1083] [Q&A]
Hello evryone, I encountered an error while running the primary input file of Fast Farm...is there someone who can tell me how to fix it and what to do? Thank you very much!
The error is the following: "forrtl: severe (157): Program Exception - access violation"

**Comment by jjonkman** [comment]
Dear @Tommi1198,

Sounds like the same issue reported in https://github.com/OpenFAST/openfast/issues/808 and solved via PR https://github.com/OpenFAST/openfast/pull/769.

Best regards,

**Comment by Tommi1198** [comment]
Sorry but I cannot understand what to do to solve my problem, since I cannot found the solution for my type of error

**Comment by jjonkman** [comment]
Dear @Tommi1198,

Making the small code change noted in PR #769 and recompiling FAST.Farm should resolve the issue.  This PR has also been merged into the main branch of OpenFAST, so, upgrading to the FAST.Farm available in OpenFAST v3.1 should also resolve the issue.

Best regards,

**Comment by andrew-platt** [comment]
Hi @Tommi1198,

Which version of FAST.Farm are you running? In addition to the issue @jjonkman mentioned, there may be another possibility depending on which version you are using.

Regards

**Comment by Tommi1198** [comment]
Dear @jjonkman , I tried to change the code noted in PR #769 , but I still have the same kind of error: "forrtl: severe (157): Program Exception - access violation"
Dear @andrew-platt , my version is OpenFAST v3.0.0, since my professor told me to use that one. 
How can I fix this error?

**Comment by Tommi1198** [comment]
Dear @jjonkman , the change that you refer to is only in IfW_FFWind_Base.F90 right? Only a line of change with respect to the original one? Am I correct?

**Comment by jjonkman** [comment]
Dear @Tommi1198,

Yes, that is correct.   Can you share the full set of messages written to the command window after executing FAST.Farm?

Best regards,

**Comment by Tommi1198** [comment]
Dear @jjonkman , here what you asked me:

FAST.Farm

Copyright (C) 2021 National Renewable Energy Laboratory
Copyright (C) 2021 Envision Energy USA LTD

This program is licensed under Apache License Version 2.0 and comes with ABSOLUTELY NO WARRANTY.
See the "LICENSE" file distributed with this software for details.
**************************************************************************************************

FAST.Farm-v3.0.0
Compile Info:
 - Compiler: Intel(R) Fortran Compiler 1900
 - Architecture: 64 bit
 - Precision: single
 - OpenMP: No
 - Date: Jun 23 2021
 - Time: 21:41:28
Execution Info:
 - Date: 03/17/2022
 - Time: 14:17:59+0100

 Heading of the FAST.Farm input file:
     Sample FAST.Farm input file
Running AWAE.
Running InflowWind.
forrtl: severe (157): Program Exception - access violation
Image              PC                Routine            Line        Source

**Comment by Tommi1198** [comment]
Dear @jjonkman , this problem is shown if I want to use the supercontroller, maybe because I have not the SuperController file...where can I find it?
I try to run fast farm without supercontroller and another type of error came out:

FAST.Farm-v3.0.0
Compile Info:
 - Compiler: Intel(R) Fortran Compiler 1900
 - Architecture: 64 bit
 - Precision: single
 - OpenMP: No
 - Date: Jun 23 2021
 - Time: 21:41:28
Execution Info:
 - Date: 03/17/2022
 - Time: 14:37:22+0100

 Heading of the FAST.Farm input file:
     Sample FAST.Farm input file
Running AWAE.
Running InflowWind.
Running WakeDynamics.
Running FASTWrapper (v1.00.00, 7-Feb-2017).
Running OpenFAST.
OpenFAST input file heading:
    DTU 10MW RWT. Compativle with OpenFAST 2.4.0

Running ElastoDyn.
Running AeroDyn.
Running InflowWind.
Running ServoDyn.
Running ServoDyn Interface for Bladed Controllers (using Intel Visual Fortran for Windows, ).
Using legacy Bladed DLL interface.

Farm_Initialize:T1:Farm_InitFAST:FWrap_Init:MeshMapCreate:CreateLoadMap_L2_to_L2:CreateMapping_Pro
jectToLine2:Node 29 does not project onto any line2 element. Closest distance is 0.9917 m.



 Aborting FAST.Farm.

How can I fix it? Thanks a lot!

**Comment by jjonkman** [comment]
Dear @Tommi1198,

So, setting `UseSC` = False in the FAST.Farm input files solves the access violation issue?

A functional super controller example is provided in the following topic on our forum: https://forums.nrel.gov/t/fast-farm/1887/161.  Before enabling the super controller, however, I would ensure that FAST.Farm runs as expected without the super controller.

Regarding the CreateLoadMap_L2_to_L2 error, this error is regarding the mesh-mapping search in the OpenFAST wrapper within FAST.Farm, which is the mesh-mapping search between the AeroDyn Line2 blade load output mesh and the radial finite-difference grid that is stored on a mesh in the OpenFAST wrapper. I would guess the problem is that your radial finite-difference grid is not larger than the blade radius in AeroDyn.

Best regards,

**Comment by Tommi1198** [comment]
Dear @jjonkman , setting UseSC = False in the FAST.Farm input files solves the access violation issue.
I now tried to run fasr farm without SC and fixing the radial finite-difference grid dimension, but another error comes out:

```
T1:FARM_InitialCO:FWrap_t0:FAST_Solution0:CalcOutputs_And_SolveForInputs:SolveOption2:SolveOption2
c_Inp2AD_SrvD:InflowWind_CalcOutput:CalcOutput:IfW_4Dext_CalcOutput [position=(-2.9327, -4.1561,
166.32) in wind-file coordinates]:Interp4D:Outside the grid bounds.
SolveOption2:SrvD_CalcOutput:DLL_controller_call:
------------------------------------------------------------------------------
Running ROSCO-HEAD-HASH-NOTFOUND
A wind turbine controller framework for public use in the scientific field
Developed in collaboration: National Renewable Energy Laboratory
                            Delft University of Technology, The Netherlands
------------------------------------------------------------------------------



 Aborting FAST.Farm.
```
What does it mean? How can I fix it?
Thank you!

**Comment by Tommi1198** [comment]
And I encountered quite the same type of error running fast farm with the turbine file without control:

T1:FARM_InitialCO:FWrap_t0:FAST_Solution0:CalcOutputs_And_SolveForInputs:SolveOption2:SolveOption2
c_Inp2AD_SrvD:InflowWind_CalcOutput:CalcOutput:IfW_4Dext_CalcOutput [position=(-2.9169,
8.75978E-15, 166.5) in wind-file coordinates]:Interp4D:Outside the grid bounds.



 Aborting FAST.Farm.

**Comment by jjonkman** [comment]
Dear @Tommi1198,

An "outside the grid bounds" error is generated when an analysis node is requesting wind data from outside of the wind domain.  This could be caused by an aerodynamic analysis node in AeroDyn that extends beyond the boundaries of the high-resolution domain, or the use of low- or high-resolution domains that extend beyond the wind domain specified within InflowWind (when `Mod_AmbWind` = 2 or 3 is selected in FAST.Farm).  I'm guessing you have not defined the low- or high-resolution domains properly within FAST.Farm.  See the "Modeling Guidance" section in the online FAST.Farm documentation for more information: https://openfast.readthedocs.io/en/main/source/user/fast.farm/ModelGuidance.html.

Best regards,

**Comment by Tommi1198** [comment]
Dear @jjonkman, I tried to re-arranged a bit the resolutions, but this error came out:

FAST.Farm

Copyright (C) 2021 National Renewable Energy Laboratory
Copyright (C) 2021 Envision Energy USA LTD

This program is licensed under Apache License Version 2.0 and comes with ABSOLUTELY NO WARRANTY.
See the "LICENSE" file distributed with this software for details.
**************************************************************************************************

FAST.Farm-v3.0.0
Compile Info:
 - Compiler: Intel(R) Fortran Compiler 1900
 - Architecture: 64 bit
 - Precision: single
 - OpenMP: No
 - Date: Jun 23 2021
 - Time: 21:41:28
Execution Info:
 - Date: 03/19/2022
 - Time: 10:01:43+0100

 Heading of the FAST.Farm input file:
     Sample FAST.Farm input file
Running AWAE.
Running InflowWind.
Running WakeDynamics.
Running FASTWrapper (v1.00.00, 7-Feb-2017).
Running OpenFAST.
OpenFAST input file heading:
    DTU 10MW RWT. Compativle with OpenFAST 2.4.0

Running ElastoDyn.
Running AeroDyn.
Running InflowWind.
Running ServoDyn.
Running ServoDyn Interface for Bladed Controllers (using Intel Visual Fortran for Windows, ).
Using legacy Bladed DLL interface.
 Time: 0 of 300 seconds.
The BEM solution is being turned off due to low TSR.  (TSR = 0). This warning will not be
repeated though the condition may persist. (See GeomPhi output channel.)

T1:FARM_InitialCO:FWrap_t0:FAST_Solution0:CalcOutputs_And_SolveForInputs:SolveOption2:SolveOption2
c_Inp2AD_SrvD:InflowWind_CalcOutput:CalcOutput:IfW_4Dext_CalcOutput [position=(-7.073, 0, 119) in
wind-file coordinates]:Interp4D:Outside the grid bounds.
SolveOption2:SrvD_CalcOutput:DLL_controller_call:
------------------------------------------------------------------------------
Running ROSCO-HEAD-HASH-NOTFOUND
A wind turbine controller framework for public use in the scientific field
Developed in collaboration: National Renewable Energy Laboratory
                            Delft University of Technology, The Netherlands
------------------------------------------------------------------------------
FWrap_t0:FWrap_CalcOutput:Nacelle-yaw error is undefined because the rotor-disk-averaged relative
wind speed is directed vertically



 Aborting FAST.Farm.

How can I fix it?

**Comment by Tommi1198** [comment]
Dear @jjonkman , I tried also to run a simulation using as example of turbine the NREL5MW Onshore, but the following error occured and I don't know how to fix it:

Farm_Initialize:For efficiency, NumPlanes has been reduced to the number of time steps (31).
Farm_Initialize:T1:Farm_InitFAST:FWrap_Init:FAST_InitializeAll:ED_Init:ED_ReadInput:ReadPrimaryFil
e:Invalid numerical input for file
"C:\Users\berna\Desktop\r-test-main\glue-codes\fast-farm\LESinflow\NRELOffshrBsline5MW_Onshore_Ela
stoDyn_8mps.dat" occurred while trying to read Gravity.



 Aborting FAST.Farm.

**Comment by jjonkman** [comment]
Dear @Tommi1198,

I wouldn't worry about solving the nacelle-yaw error issue until you solve the "outside the grid bounds" issue.

The error regarding Gravity is triggered because of an improper formatting of your input file(s).  My guess is you upgraded your version of OpenFAST and FAST.Farm without upgrading your input file(s).  See the documentation here for a description of the input file changes with each version: https://openfast.readthedocs.io/en/main/source/user/api_change.html.  Gravity was removed from the ElastoDyn input file in OpenFAST v3.1.

Best regards,

**Comment by Tianyao314** [comment]
dear @jjonkman 

I've set UseSC = False in the FAST.Farm input file, but I still get this error. Is there any other way to solve this error?

![error](https://github.com/OpenFAST/openfast/assets/55082013/2a3b9e52-a0ce-4cb3-82cc-d4994b058d6b)

Best regards,

**Reply by jjonkman** [reply]
Dear @Tianyao314,

These errors appear to be from an incompatibility between the wake finite-difference grid you've defined in FAST.Farm and the aerodynamic nodes in AeroDyn.  I'd have to see your FAST.Farm and OpenFAST input files to understand what specifically the issue is.

Best regards,

**Reply by Tianyao314** [reply]
Dear @jjonkman 

Thanks for your prompt, I have changed the Settings of the grid in the FAST.Farm input file, and now it can run normally.

Best Regards,
